### PR TITLE
Modernize Ars component

### PR DESCRIPTION
**Pull request by nhunzaker**
This commit updates the Ars component to use React.Component, it also adds Flow definitions and eliminates the need for the Sync mixin for this component.

In the process of doing this, I hit a few snags using `mixed` as the type for record ids, which turned into a bit of a large Flow type update.

### CODE REVIEW

File: src/components/ars.js [hunk 1]
```diff
@@ -1,105 +1,132 @@
...
+  multiselect: boolean,
+  resource: string,
+  mode: 'gallery' | 'table',
+  picked: string | number | Array<string | number>,
+  columns?: string[],
+  url: string,
+  onFetch: (*) => Record,
+  onError: (*) => *,
+  onChange: (*) => any
+}
```
Comment (nhunzaker) [inline src/components/ars.js]: I don't think all of these attributes are necessary, but I need them temporarily until I completely get rid of the mixins for fetching data.

File: src/components/multiselection.js [hunk 2]
```diff
@@ -11,27 +11,27 @@ import SelectionText from './selection-text'
 type Props = {
   onClick: Event => *,
   resource: string,
-  slug: string[],
+  slugs: Array<string | number>,
```
Comment (dbernheisel) [inline src/components/multiselection.js]: 🎊

File: src/components/ars.js [hunk 3]
```diff
@@ -1,105 +1,132 @@
...
 
   getPicker() {
     let { picked } = this.state
     let { columns, multiselect, mode } = this.props
 
     return (
       <Picker
-        key="dialog"
-        ref="picker"
         {...this.syncProps()}
```
Comment (efatsi) [inline src/components/ars.js]: out of curiosity, where was `syncProps()` defined before this PR? looks like it was added here but this call existed before

File: src/components/ars.js [hunk 4]
```diff
@@ -1,105 +1,132 @@
...
   _onOpenClick() {
     this.setState({ dialogOpen: true })
-  },
+  }
 
-  _onGalleryPicked(picked) {
-    this.setState({ picked }, this._triggerChange)
-  },
+  _onGalleryPicked(picked: *) {
+    this.setState({ picked }, this._triggerChange.bind(this))
```
Comment (dbernheisel) [inline src/components/ars.js]: just curious, what do you think of rewriting these callbacks as instance functions so you avoid `bind(this)`. Doesn't really matter, but I found a preference in that style.

```js
_triggerChange = () => {}
\\..
this.setState({ picked }), this._triggerChange)
```

File: src/components/ars.js (hunk 3 above)
Comment (dbernheisel) [reply src/components/ars.js]: it came from `mixins: [Sync]`

Comment (nhunzaker) [reply src/components/ars.js]: It was defined in the Sync mixin:
https://github.com/vigetlabs/ars-arsenal/blob/master/src/mixins/sync.js

Long term, I think the best plan for this sort of thing is to expose a configuration object using context. That'll avoid needing to pass around a bunch of props everywhere that aren't being used. 

That's coming up pretty soon!

File: src/components/ars.js (hunk 4 above)
Comment (nhunzaker) [reply src/components/ars.js]: Thanks for the prod. I think this is a great idea. I'm just about done moving everything over to the updated React API. Do you mind if I do this in one swoop?

**Comment by nhunzaker** [comment]
Thanks, everyone!

**Pull request merged**